### PR TITLE
StaticKeyMaterialByronPbft:  encapsulate the local key material necessary to run RealPBFT

### PR DESCRIPTION
**Pull request by deepfire**
The name is specifically including `Pbft`, because Byron Classic
also required a VSS keypair, which the PBFT variant does not.

Alternatively, this type could have been placed in the PBFT protocol,
under a different name, but it would have been even more unappropriately
over-specific there.

Ideal(-istic)ally, the notion of a per-protocol/block/ledger type
for node-specific key material could have been appropriately indexed
by the protocol/block types, but as yet there's no clear picture
of tradeoffs involved, since there's just a single "real" implementation
so far.

### CODE REVIEW

**Comment by dcoutts** [comment]
Those build system cleanups are independent of the code changes. They should not be in the same PR.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 1]
```diff
@@ -165,6 +167,22 @@ instance (ByronGiven, Typeable cfg) => Measured BlockMeasure (ByronBlock cfg) wh
...
+{-------------------------------------------------------------------------------
+  Secrets
+-------------------------------------------------------------------------------}
+
+-- This is specifically including Pbft in the name, because Byron Classic
+-- also required a VSS keypair, which Pbft does not.
+--
+-- Alternatively, this type could have been placed in the PBFT protocol,
+-- under a different name, but it would have been even more unappropriately
+-- over-specific there.
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: I'm not sure what to make of this comment. It's not API documentation. It seems to really be a commit message or PR review message suggesting alternative ways of doing it.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 2]
```diff
@@ -165,6 +167,22 @@ instance (ByronGiven, Typeable cfg) => Measured BlockMeasure (ByronBlock cfg) wh
...
+data StaticKeyMaterialByronPbft = StaticKeyMaterialByronPbft {
+      skmbpSignKey     :: Crypto.SigningKey
+      -- | Concrete, because `cardano-crypto` does not provide an abstract delegation certificate.
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: API docs should first say what it is, what it is for, and lastly why it is the way it is. This is just a cryptic comment on why something is the way it is (without saying what that alternative is).

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 3]
```diff
@@ -45,8 +45,9 @@ protocolInfoByron :: NumCoreNodes
                   -> CoreNodeId
                   -> PBftParams
                   -> Cardano.Genesis.Config
+                  -> StaticKeyMaterialByronPbft
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Make this optional.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 4]
```diff
@@ -58,9 +59,9 @@ protocolInfoByron (NumCoreNodes numCoreNodes) (CoreNodeId nid) params gc =
                   --TODO: also allow not being a BFT leader:
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: You can fix this TODO now if you make the secret key material optional.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 1 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Perhaps this type only needs to be defined in `Ouroboros.Consensus.Node.ProtocolInfo.Byron`

As for naming, perhaps `PbftLeaderCredentials`? Yes, mentioning PPBT makes sense. The inclding "byron" in the name is unnecessary, as PBFT is only for that era.

@nc6 opinions on either/both?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 2 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 1 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: @nc6, waiting for your take!

Meanwhile I've implemented the initial Duncan's suggestion:

1. renamed to `PbftLeaderCredentials`
2. moved to `Ouroboros.Consensus.Node.ProtocolInfo.Byron`

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 3 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 4 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done!

**Comment by deepfire** [comment]
> Those build system cleanups are independent of the code changes. They should not be in the same PR.

Done

**Comment by deepfire** [comment]
@dcoutts, all points addressed, waiting for @nc6's feedback on our naming choice.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 5]
```diff
@@ -87,19 +92,13 @@ protocolInfoByron (NumCoreNodes numCoreNodes) (CoreNodeId nid) params gc =
       , pInfoInitState  = ()
       }
   where
+    proofOfCredentials :: CoreNodeId -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
+    proofOfCredentials nid' (PbftLeaderCredentials sk dlg) =
+      PBftIsLeader
+      { pbftCoreNodeId = nid'
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: One of the next steps will be to compute this based on the genesis verification key (as in my original WIP/hacking branch).

**Comment by dcoutts** [comment]
Next steps for this bit of code:

* `pbftNumNodes` computed (from `numGenesisKeys`) rather than passed in
* `pbftCoreNodeId` computes rather than passed in
* switch core pbft `IsLeader` type to use the given delegation cert rather than looking it up
* get protocol and software version from external config
* `pbftSignatureWindow` is always K I think
* `pbftSignatureThreshold` is 0.22 for 7 nodes, but ought in principle to come from external config

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 5 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Good point, I had the same thought on Slack!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 6]
```diff
@@ -45,26 +55,26 @@ protocolInfoByron :: NumCoreNodes
...
             encNodeConfigP = PBftNodeConfig {
                   pbftParams   = params
-                    { pbftNumNodes = fromIntegral numCoreNodes
+                    { -- TODO (Duncan): compute (from numGenesisKeys)
+                      pbftNumNodes = fromIntegral numCoreNodes
                       -- Set the signature window to be short for the demo.
+                      -- TODO: Is it always 'k'?
                     , pbftSignatureWindow = 7
+                      -- TODO: make pbftSignatureThreshold default to a value computed from pbftNumNodes,
+                      --       but configurable?
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: `pbftSignatureThreshold` is already configurable, it's just not overridden here...

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 7]
```diff
@@ -35,6 +36,15 @@ import           Ouroboros.Consensus.Ledger.Byron.Config
 
 import qualified Test.Cardano.Chain.Genesis.Dummy as Dummy
 
+{-------------------------------------------------------------------------------
+  Credentials
+-------------------------------------------------------------------------------}
+
+data PbftLeaderCredentials = PbftLeaderCredentials {
+      plcSignKey     :: Cardano.SigningKey
```
[the remark this answers is not in the document]
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: @nc6, my apologies -- I'll start a followup PR as soon as I get `cardano-node` working.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 6 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: The idea, I think, is to supply the configuration from the outside, the same way `PbftLeaderCredentials` got added.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Also, I think that the use for the key is implied (albeit, perhaps, weakly) by the type name: `PbftLeaderCredentials` -- i.e. it seems to say something-leader-something-sign-blocks-something, I'd think.

**Pull request merged**